### DISCUSSION TITLE
forcing dhcp ?

### DISCUSSION THREAD

**Discussion by zcattacz** [micropython/micropython#12974] [ESP32]
When a network is down, there are connectivity issue at different level, wifi>dhcp>lan>wan>server
I wonder if there is a way to check dhcp validity, for example to force mpy to force a dhcp lease request/update or so without reconnect wifi?

**Comment by karfas** [comment]
I don't see the point here.

Renewing the DHCP lease (without WLAN disconnect) will most likely not fix an arbitrary connection problem between your device and a server somewhere in the "cloud".

A complete reconnect, on the other hand, might be able to find an access point on another network (with working internet access).

**Reply by zcattacz** [reply]
It happens to me, when WiFi is indeed having issues, either wifi channel change, or bad signal, a toggle/re-scan is required, both takes quite a few seconds on my ESP32 board, and they are blocking calls. 
I need to avoid blocking calls unless necessary to minimize interference with other coros jobs.

So to work out the issue, e.g. if the dhcp works, the lan works, there no need to mess with wifi. 
The wan gateway may be down, or the route to mqtt server maybe down, the client just need to keep pinging the gw/srv untill they got up and retry.

Currently I use ping, I just wonder if dhcp can be triggered manually, since the precondition of succ ping is the interface have a valid IP, dhcp doesn't have this limitation. And if there is a IP issue, it's likely dhcp refresh will fix it. 
This can serve as a failover measurement between gw ping and wifi toggle.

**Reply by karfas** [reply]
`WLAN.connect()` **does not** block on the ESP32; a simple wrapper (see e.g. https://github.com/karfas/upy-simple-app/blob/master/lib/asy_wlan_client.py ) can be used to connect using an `async` function.
However, in the moment no non-blocking version of `WLAN.scan()` exists.

**Reply by zcattacz** [reply]
Once, I found the ESP32 in a weird state, connect() call is not blocking, but it does not connect either. I always toggle wifi at 1s interval before connect, this didn't make any changes that time. Only after a scan, it immediately start to connect as usual. (maybe some outdated AP info? I am not sure), so now I also always scan before connect. unless configured AP is hidden. I figure if I can minimize wifi recovery, extra steps to make sure wifi connects in all scenario is affordable.

**Reply by karfas** [reply]
OK, but I would suspect that the required(?) scan in your case has nothing to do with DHCP and should be fixed on a (I suppose) lower level than micropython.

In case you have means to verify the behaviour, you might want to try recent micropython versions using one of the newer (v5.x) ESP-IDF releases.

It would also be interesting what status the connect() call returns in that case (with reconnects=0 or similar).

**Reply by karfas** [reply]
A quick search gave me https://github.com/orgs/micropython/discussions/9828#discussioncomment-5053505 - this uses an extra thread to create a non-blocking, async scan.

**Comment by sophiedegran** [comment]
I'm not sure if working on DHCP LEASE will solve your problem of wifi instability... (for me certainly not)
The LEASE TIME is defined in dhcpserver.c (24 * 60 * 60) in sec
Have a look ☺

**Reply by zcattacz** [reply]
Thank you, will take a look

**Reply by sophiedegran** [reply]
Sorry but I suppose you was on AP mode so that will not correspond.. The idea to do async connection on STA_IF can be good if all your project is on async mode, did your try karfas example?
